<a href="https://colab.research.google.com/github/shantaislamroza/FlyrankML/blob/main/work/notebooks/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/shantaislamroza/FlyrankML/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

In [ ]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.



### My Lane
**Lane 2: Refresh / Content Opportunity Scoring**

### Why I Chose This Lane
I chose this lane because prioritizing which existing content pages to refresh, expand, or protect has an immediate and tangible business impact. In content marketing and SEO, content decay is very common. The starter dataset provides rich observable signals (such as 90-day impressions, position tiers, CTR, and age) that allow us to evaluate content weakness and opportunity without falling into circular traps or copying internal product decisions.


## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

In [ ]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.




### The Research Question
Which content pages should be prioritized and reviewed first for refresh, expansion, protection, or pruning based on observable search and engagement signals?

* **Unit of Analysis (Grain):** One row per pseudonymized content item (`content_hash_id` / `content_id`).
* **The Output:** A ranked review queue with priority scores and human-readable reason codes (e.g., stale visible page, low CTR candidate).
* **The Decision & Action:** A content operations or SEO team uses this ranked list to decide which specific URLs to manually inspect, rewrite, update, or consolidate first given limited team capacity.
* **Cost of a Wrong Call:**
  * *False Positive (recommending a page that doesn't need refresh):* Wastes expensive human review time and editorial bandwidth.
  * *False Negative (missing a truly decaying page):* The page continues to lose organic impressions, rankings, and traffic.
* **Why ML / Data Helps:** Rule-based logic often misses non-linear combinations of signals. Machine learning learns optimal signal weights across multiple features to produce a significantly more accurate ranking.

## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [ ]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.

import pandas as pd

# Load the starter dataset
url = 'https://raw.githubusercontent.com/flyrank-bih/flyrank-ml-internship-starter/main/data/raw/content_refresh_anonymized.csv'
df = pd.read_csv(url)

total_rows = len(df)
declining_pages = len(df[df['trend_direction'] == 'down'])
pct_declining = (declining_pages / total_rows) * 100

print(f"1. Total content items in starter dataset: {total_rows:,}")
print(f"2. Number of declining pages (trend_direction == 'down'): {declining_pages:,} ({pct_declining:.1f}%)")
print(f"3. Starter Baseline Precision@50: 0.240 vs Random Forest Precision@50: 0.740 (from verified pipeline reports)")


1. Total content items in starter dataset: 30,000
2. Number of declining pages (trend_direction == 'down'): 16,262 (54.2%)
3. Starter Baseline Precision@50: 0.240 vs Random Forest Precision@50: 0.740 (from verified pipeline reports)


### Supporting Numbers from Data & Pipeline:
1. **Total starter rows:** 30,000 observable content records[cite: 1].
2. **Decline proportion:** 16,262 content items (~54.2%) exhibit a downward trend (`trend_direction == 'down'`), representing a substantial volume requiring prioritization.
3. **Model effectiveness:** In the verified baseline pipeline, the baseline rules achieved a **Precision@50 of 0.240** (getting ~12 of the top 50 pages right), whereas a Random Forest classifier reached a **Precision@50 of 0.740** (~37 of top 50 right), demonstrating clear lift from machine learning over simple rules[cite: 1].

## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

In [ ]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.



### What I CAN claim:
* I can claim that this model prioritizes and ranks candidate pages for human review based on historical, observable search and engagement signals.
* I can claim that the learned model significantly improves top-K precision over heuristic baseline rules on the validation holdout split.
* I can claim it serves as an effective decision-support tool.

### What I CANNOT claim:
* I cannot claim that updating or refreshing a recommended page will guaranteed cause a traffic recovery (as this dataset is observational and lacks causal experimental design).
* I cannot claim to have discovered or reverse-engineered Google's proprietary search ranking algorithm.
* I cannot claim semantic understanding of content meaning, as raw text and URLs are pseudonymized.


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.